diff --git a/server/build_without_gradle.sh b/server/build_without_gradle.sh
index 0f86c29..c3c6630 100755
--- a/server/build_without_gradle.sh
+++ b/server/build_without_gradle.sh
@@ -57,7 +57,7 @@ javac -bootclasspath "$ANDROID_JAR" -cp "$CLASSES_DIR" -d "$CLASSES_DIR" \
 echo "Dexing..."
 cd "$CLASSES_DIR"
 
-if [[ $PLATFORM_VERSION -lt 31 ]]
+if [[ $PLATFORM -lt 31 ]]
 then
     # use dx
     "$ANDROID_HOME/build-tools/$BUILD_TOOLS/dx" --dex \
